source/agora/consensus/validation/Block.d

```diff
@@ -136,7 +140,34 @@ public string isInvalidReason (const ref Block block, Height prev_height,
             return fail_reason;
     }
 
-    return null;
+    return validateBlockTimestamp(prev_timestamp, block.header.timestamp, curr_timestamp, blockTimestampToleranceSecs);
+}
+
+static public string validateBlockTimestamp(ulong prev_block_ts, ulong new_block_ts, ulong curr_timestamp,
+        uint blockTimestampToleranceSecs) @safe nothrow
+{
+    string res;
+    try
+    {
+        import std.conv : to;
+        import std.datetime.systime;
+        if (new_block_ts <= prev_block_ts)
+            res = "the timestamp in consensus data is not greater than the timestamp in the last block\n" ~
+                    "new block timestamp: " ~ to!string(SysTime(unixTimeToStdTime(new_block_ts))) ~ "\n" ~
+                    "previous block's timestamp: " ~ to!string(SysTime(unixTimeToStdTime(prev_block_ts)));
+        else if (new_block_ts > curr_timestamp + blockTimestampToleranceSecs)
+            res =  "(the timestamp in consensus data) is greater than (the timestamp in the last block + tolerance)\n" ~
```

Comment:
Again this doesn't really make sense. If the nodes are late to nominate then they will not be able to recover.

Reply:
see https://github.com/bpfkorea/agora/pull/1437#discussion_r546591387

if you mean the node's current clock is too back in the past, so it will not accept anything: I think we already have some assumption about how close the clocks of the nodes have to be together. Why was the original clock synchronization algorithm added? Where else do we need the clocks to be synchronized and what happen there if they are not?

Reply:
> Why was the original clock synchronization algorithm added?

In this case the node itself may reject voting on a nomination, but if 51% of the network agrees on the vote then it will go through anyway.

In your case you're adding this check to the block validation itself. That's different. Because it means no block will ever be accepted if it exceeds the tolerance. But what do we set the tolerance to? 1 hour? 1 day?

What do other blockchains do? Is there a protocol-level rule like this?

Reply:
1. let's say the node X clock's is really behind and 51% of the nodes accepted a nomination and the block got externalized
2. when the block reaches node X it will find it invalid because of the tolerance and it will exit as an externalized invalid block was received(not ideal, but we can use different tolerance value for voting and accepting blocks)
3. the node restarts next day
4. now the node's current time + tolerance will accept the block that was previously rejected

so the criteria of accepting the block is less stringent as the physical time goes on

Reply:
Alright. Let's just get some test-cases written for this. It's better to confirm our expectations.